components/supervisor/pkg/supervisor/git.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"context"
+	"time"
+
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/gitpod"
+)
+
+// GitTokenProvider provides tokens from git hosting services
+type GitTokenProvider struct {
+	gitpodAPI gitpod.APIInterface
+}
+
+// NewGitTokenProvider creates a new instance of gitTokenProvider
+func NewGitTokenProvider(gitpodAPI gitpod.APIInterface) *GitTokenProvider {
+	return &GitTokenProvider{
+		gitpodAPI: gitpodAPI,
+	}
+}
+
+// GetToken resolves a token from a git hosting service
+func (p *GitTokenProvider) GetToken(ctx context.Context, req *api.GetTokenRequest) (tkn *Token, err error) {
+	token, err := p.gitpodAPI.GetToken(ctx, &gitpod.GetTokenSearchOptions{
```

Comment:
What if gitpod api is nil?

Reply:
Seems like in that case the whole supervisor wouldn't work properly and it is ok to fail. I don't see how that could happen though.

Reply:
In other cases we have guards.